File: app/src/main/java/com/gmail/mostafa/ma/saleh/yawa/retrofit/RetrofitManager.java
package com.gmail.mostafa.ma.saleh.yawa.retrofit;

import android.support.annotation.NonNull;

import com.gmail.mostafa.ma.saleh.yawa.models.Day;
import com.gmail.mostafa.ma.saleh.yawa.models.ForecastResponse;
import com.gmail.mostafa.ma.saleh.yawa.utilities.Constants;

import java.io.IOException;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitManager {

    private static RetrofitManager INSTANCE = new RetrofitManager();
    private Requests requests;

    private RetrofitManager() {
        buildClient();
    }

    public static RetrofitManager getInstance() {
        return INSTANCE;
    }

    private void buildClient() {
        HttpLoggingInterceptor logger = new HttpLoggingInterceptor();
        logger.setLevel(HttpLoggingInterceptor.Level.BODY);
        Interceptor appIdInterceptor = new Interceptor() {
            @Override
            public okhttp3.Response intercept(Chain chain) throws IOException {
                Request original = chain.request();
                HttpUrl.Builder url = original.url().newBuilder()
                        .addQueryParameter("APPID", Constants.APP_ID);
                Request request = original.newBuilder()
                        .url(url.build()).build();
                return chain.proceed(request);
            }
        };
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(logger)
                .addInterceptor(appIdInterceptor)
                .build();
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(Constants.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();
        requests = retrofit.create(Requests.class);
    }

    public void getForecast(String cityId, String tempUnit, final OnFinishedListener<Day[]> onFinishedListener) {
        requests.getForecast(cityId, tempUnit).enqueue(new Callback<ForecastResponse>() {
            @Override
            public void onResponse(@NonNull Call<ForecastResponse> call, @NonNull Response<ForecastResponse> response) {
                ForecastResponse forecastResponse = response.body();
                if (forecastResponse != null) {
                    onFinishedListener.onComplete(forecastResponse.days, null);
                }
            }

            @Override
            public void onFailure(@NonNull Call<ForecastResponse> call, @NonNull Throwable t) {
                onFinishedListener.onComplete(null, t.getLocalizedMessage());
            }
        });
    }

}

File: app/src/main/java/com/gmail/mostafa/ma/saleh/yawa/retrofit/RetrofitManager.kt
package com.gmail.mostafa.ma.saleh.yawa.retrofit

import com.gmail.mostafa.ma.saleh.yawa.models.Day
import com.gmail.mostafa.ma.saleh.yawa.models.ForecastResponse
import com.gmail.mostafa.ma.saleh.yawa.utilities.Constants
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitManager {
    private val requests by lazy { buildClient().create(Requests::class.java) }

    private fun buildClient(): Retrofit {
        val logger = HttpLoggingInterceptor()
        logger.level = HttpLoggingInterceptor.Level.BODY
        val client = OkHttpClient.Builder()
                .addInterceptor(logger)
                .addInterceptor { chain ->
                    val original = chain.request()
                    val url = original.url().newBuilder()
                            .addQueryParameter("APPID", Constants.APP_ID)
                    val request = original.newBuilder()
                            .url(url.build()).build()
                    chain.proceed(request)
                }
                .build()
        return Retrofit.Builder()
                .baseUrl(Constants.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build()
    }

    fun getForecast(cityId: String, tempUnit: String, onFinishedListener: OnFinishedListener<List<Day>>) {
        requests.getForecast(cityId, tempUnit).enqueue(object : Callback<ForecastResponse> {
            override fun onResponse(call: Call<ForecastResponse>, response: Response<ForecastResponse>) {
                val forecastResponse = response.body()
                onFinishedListener.onComplete(forecastResponse?.days, null)
            }

            override fun onFailure(call: Call<ForecastResponse>, t: Throwable) {
                onFinishedListener.onComplete(null, t.localizedMessage)
            }
        })
    }
}
